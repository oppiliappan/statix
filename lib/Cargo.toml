[package]
name = "lib"
version = "0.0.0"
edition = "2018"
license = "MIT"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rnix = "0.12.0"
if_chain = "1.0"
macros = { path = "../macros" }
lazy_static = "1.5"
rowan = "0.15.17"
serde_json = { version = "1.0.99", optional = true }

[dependencies.serde]
version = "1.0.130"
features = [ "derive" ]
optional = true

[features]
default = []
json-out = [ "serde", "serde_json" ]
