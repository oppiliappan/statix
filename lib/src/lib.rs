#![recursion_limit = "1024"]
mod lints;
mod make;
mod parse;
pub mod session;
mod utils;

pub use lints::LINTS;
use session::SessionInfo;

use rnix::{parser::ParseError, SyntaxElement, SyntaxKind, TextRange};
use std::{convert::Into, default::Default};

#[cfg(feature = "json-out")]
use serde::{
    ser::{SerializeStruct, Serializer},
    Serialize,
};

#[derive(Debug, Default)]
#[cfg_attr(feature = "json-out", derive(Serialize))]
pub enum Severity {
    #[default]
    Warn,
    Error,
    Hint,
}

/// Report generated by a lint
#[derive(Debug, Default)]
#[cfg_attr(feature = "json-out", derive(Serialize))]
pub struct Report {
    /// General information about this lint and where it applies.
    pub note: &'static str,
    /// An error code to uniquely identify this lint
    pub code: u32,
    /// Report severity level
    pub severity: Severity,
    /// Collection of diagnostics raised by this lint
    pub diagnostics: Vec<Diagnostic>,
}

impl Report {
    /// Construct a report. Do not invoke `Report::new` manually, see `lint` macro
    #[must_use]
    pub fn new(note: &'static str, code: u32) -> Self {
        Self {
            note,
            code,
            ..Default::default()
        }
    }
    /// Add a diagnostic to this report
    #[allow(clippy::return_self_not_must_use)]
    pub fn diagnostic<S: AsRef<str>>(mut self, at: TextRange, message: S) -> Self {
        self.diagnostics.push(Diagnostic::new(at, message));
        self
    }
    /// Add a diagnostic with a fix to this report
    #[allow(clippy::return_self_not_must_use)]
    pub fn suggest<S: AsRef<str>>(
        mut self,
        at: TextRange,
        message: S,
        suggestion: Suggestion,
    ) -> Self {
        self.diagnostics
            .push(Diagnostic::suggest(at, message, suggestion));
        self
    }
    /// Set severity level
    #[allow(clippy::return_self_not_must_use, clippy::must_use_candidate)]
    pub fn severity(mut self, severity: Severity) -> Self {
        self.severity = severity;
        self
    }
    /// A range that encompasses all the suggestions provided in this report
    pub fn total_suggestion_range(&self) -> Option<TextRange> {
        self.diagnostics
            .iter()
            .filter_map(|d| Some(d.suggestion.as_ref()?.at))
            .reduce(rnix::TextRange::cover)
    }
    /// A range that encompasses all the diagnostics provided in this report
    pub fn total_diagnostic_range(&self) -> Option<TextRange> {
        self.diagnostics
            .iter()
            .map(|d| d.at)
            .reduce(rnix::TextRange::cover)
    }
    /// Unsafe but handy replacement for above
    #[must_use]
    pub fn range(&self) -> TextRange {
        self.total_suggestion_range().unwrap()
    }
    /// Apply all diagnostics. Assumption: diagnostics do not overlap
    pub fn apply(&self, src: &mut String) {
        for d in &self.diagnostics {
            d.apply(src);
        }
    }
    /// Create a report out of a parse error
    #[must_use]
    pub fn from_parse_err(err: &ParseError) -> Self {
        let at = match err {
            ParseError::Unexpected(at)
            | ParseError::UnexpectedExtra(at)
            | ParseError::UnexpectedWanted(_, at, _)
            | ParseError::UnexpectedDoubleBind(at)
            | ParseError::DuplicatedArgs(at, _) => at,
            ParseError::UnexpectedEOF | ParseError::UnexpectedEOFWanted(_) => {
                &TextRange::empty(0u32.into())
            }
            _ => panic!("report a bug, pepper forgot to handle a parse error"),
        };
        let mut message = err.to_string();
        message
            .as_mut_str()
            .get_mut(0..1)
            .unwrap()
            .make_ascii_uppercase();
        Self::new("Syntax error", 0)
            .diagnostic(*at, message)
            .severity(Severity::Error)
    }
}

/// Mapping from a bytespan to an error message.
/// Can optionally suggest a fix.
#[derive(Debug)]
pub struct Diagnostic {
    pub at: TextRange,
    pub message: String,
    pub suggestion: Option<Suggestion>,
}

impl Diagnostic {
    /// Construct a diagnostic.
    pub fn new<S: AsRef<str>>(at: TextRange, message: S) -> Self {
        Self {
            at,
            message: message.as_ref().into(),
            suggestion: None,
        }
    }
    /// Construct a diagnostic with a fix.
    pub fn suggest<S: AsRef<str>>(at: TextRange, message: S, suggestion: Suggestion) -> Self {
        Self {
            at,
            message: message.as_ref().into(),
            suggestion: Some(suggestion),
        }
    }
    /// Apply a diagnostic to a source file
    pub fn apply(&self, src: &mut String) {
        if let Some(s) = &self.suggestion {
            s.apply(src);
        }
    }
}

#[cfg(feature = "json-out")]
impl Serialize for Diagnostic {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("Diagnostic", 3)?;
        let at = {
            let start = usize::from(self.at.start());
            let end = usize::from(self.at.end());
            (start, end)
        };
        s.serialize_field("at", &at)?;
        s.serialize_field("message", &self.message)?;
        if let Some(suggestion) = &self.suggestion {
            s.serialize_field("suggestion", suggestion)?;
        }
        s.end()
    }
}

/// Suggested fix for a diagnostic, the fix is provided as a syntax element.
/// Look at `make.rs` to construct fixes.
#[derive(Debug)]
pub struct Suggestion {
    pub at: TextRange,
    pub fix: SyntaxElement,
}

impl Suggestion {
    /// Construct a suggestion.
    pub fn new<E: Into<SyntaxElement>>(at: TextRange, fix: E) -> Self {
        Self {
            at,
            fix: fix.into(),
        }
    }
    /// Apply a suggestion to a source file
    pub fn apply(&self, src: &mut String) {
        let start = usize::from(self.at.start());
        let end = usize::from(self.at.end());
        src.replace_range(start..end, &self.fix.to_string());
    }
}

unsafe impl Send for Suggestion {}

#[cfg(feature = "json-out")]
impl Serialize for Suggestion {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut s = serializer.serialize_struct("Suggestion", 2)?;
        let at = {
            let start = usize::from(self.at.start());
            let end = usize::from(self.at.end());
            (start, end)
        };
        let fix = self.fix.to_string();
        s.serialize_field("at", &at)?;
        s.serialize_field("fix", &fix)?;
        s.end()
    }
}

/// Lint logic is defined via this trait. Do not implement manually,
/// look at the `lint` attribute macro instead for implementing rules
pub trait Rule {
    fn validate(&self, node: &SyntaxElement, sess: &SessionInfo) -> Option<Report>;
}

/// Contains information about the lint itself. Do not implement manually,
/// look at the `lint` attribute macro instead for implementing rules
pub trait Metadata {
    fn name(&self) -> &'static str;
    fn note(&self) -> &'static str;
    fn code(&self) -> u32;
    fn report(&self) -> Report;
    fn match_with(&self, with: &SyntaxKind) -> bool;
    fn match_kind(&self) -> Vec<SyntaxKind>;
}

/// Contains offline explanation for each lint
/// The `lint` macro scans nearby doc comments for
/// explanations and derives this trait.
///
/// FIXME: the lint macro does way too much, maybe
/// split it into smaller macros.
pub trait Explain {
    fn explanation(&self) -> &'static str {
        "no explanation found"
    }
}

/// Combines Rule and Metadata, do not implement manually, this is derived by
/// the `lint` macro.
pub trait Lint: Metadata + Explain + Rule + Send + Sync {}

/// Helper utility to take lints from modules and insert them into a map for efficient
/// access. Mapping is from a `SyntaxKind` to a list of lints that apply on that Kind.
///
/// See `lints.rs` for usage.
#[macro_export]
macro_rules! lints {
    ($($s:ident),*,) => {
        lints!($($s),*);
    };
    ($($s:ident),*) => {
        $(
            mod $s;
        )*
        ::lazy_static::lazy_static! {
            pub static ref LINTS: Vec<&'static Box<dyn $crate::Lint>> = {
                let mut v = Vec::new();
                $(
                    {
                        let temp_lint = &*$s::LINT;
                        v.push(temp_lint);
                    }
                )*
                v
            };
        }
    }
}
