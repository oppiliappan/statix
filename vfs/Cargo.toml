[package]
name = "vfs"
version = "0.0.0"
edition = "2018"
license = "MIT"

[dependencies]
indexmap = "2.10.0"
rayon = "1.10.0"
