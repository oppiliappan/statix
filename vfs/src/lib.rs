use std::{
    collections::HashMap,
    default::Default,
    path::{Path, PathBuf},
};

use indexmap::IndexSet;
use rayon::prelude::*;

#[derive(Copy, Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct FileId(pub usize);

#[derive(Debug, Default)]
pub struct Interner {
    map: IndexSet<PathBuf>,
}

impl Interner {
    pub fn get<P: AsRef<Path>>(&self, path: P) -> Option<FileId> {
        self.map.get_index_of(path.as_ref()).map(FileId)
    }
    pub fn intern(&mut self, path: PathBuf) -> FileId {
        let (id, _) = self.map.insert_full(path);
        FileId(id)
    }
    pub fn lookup(&self, file: FileId) -> Option<&Path> {
        self.map.get_index(file.0).map(PathBuf::as_path)
    }
}

#[derive(Default)]
pub struct ReadOnlyVfs {
    interner: Interner,
    data: HashMap<FileId, Vec<u8>>,
}

impl ReadOnlyVfs {
    pub fn singleton<P: AsRef<Path>>(path: P, contents: &[u8]) -> Self {
        let mut vfs = ReadOnlyVfs::default();
        vfs.set_file_contents(path, contents);
        vfs
    }
    pub fn alloc_file_id<P: AsRef<Path>>(&mut self, path: P) -> FileId {
        self.interner.intern(path.as_ref().to_owned())
    }
    #[must_use]
    pub fn len(&self) -> usize {
        self.data.len()
    }
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }
    #[must_use]
    pub fn file_path(&self, file_id: FileId) -> &Path {
        self.interner.lookup(file_id).unwrap()
    }
    #[must_use]
    pub fn get(&self, file_id: FileId) -> &Vec<u8> {
        self.data.get(&file_id).unwrap()
    }
    #[must_use]
    pub fn get_str(&self, file_id: FileId) -> &str {
        std::str::from_utf8(self.get(file_id)).unwrap()
    }
    pub fn get_mut(&mut self, file_id: FileId) -> &mut Vec<u8> {
        self.data.get_mut(&file_id).unwrap()
    }
    pub fn set_file_contents<P: AsRef<Path>>(&mut self, path: P, contents: &[u8]) {
        let file_id = self.alloc_file_id(path);
        self.data.insert(file_id, contents.to_owned());
    }
    pub fn iter(&self) -> impl Iterator<Item = VfsEntry<'_>> {
        self.data.keys().map(move |file_id| VfsEntry {
            file_id: *file_id,
            file_path: self.file_path(*file_id),
            contents: self.get_str(*file_id),
        })
    }
    #[must_use]
    pub fn par_iter(&self) -> impl ParallelIterator<Item = VfsEntry<'_>> {
        self.data.par_iter().map(move |(file_id, _)| VfsEntry {
            file_id: *file_id,
            file_path: self.file_path(*file_id),
            contents: self.get_str(*file_id),
        })
    }
}

pub struct VfsEntry<'ρ> {
    pub file_id: FileId,
    pub file_path: &'ρ Path,
    pub contents: &'ρ str,
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn trivial() {
        let mut vfs = ReadOnlyVfs::default();
        let f1 = "a/b/c";
        let id1 = vfs.alloc_file_id(f1);
        let data = "hello".as_bytes().to_vec();
        vfs.set_file_contents(f1, &data);
        assert_eq!(vfs.get(id1), &data);
    }
}
