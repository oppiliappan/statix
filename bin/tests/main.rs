use lib::session::{SessionInfo, Version};

macro_rules! session_info {
    ($version:expr) => {{
        let v: Version = $version.parse().unwrap();
        SessionInfo::from_version(v)
    }};
}

mod util {
    #[macro_export]
    macro_rules! test_lint {
        ($tname:ident => $sess:expr, $($tail:tt)*) => {
            test_lint!($tname => $sess);
            test_lint!($($tail)*);
        };
        ($tname:ident, $($tail:tt)*) => {
                test_lint!($tname);
                test_lint!($($tail)*);
        };
        ($tname:ident) => {
            test_lint!($tname => session_info!("2.6"));
        };
        ($tname:ident => $sess:expr) => {
            #[test]
            fn $tname() {
                use statix::{config::OutFormat, traits::WriteDiagnostic, lint};
                use vfs::ReadOnlyVfs;

                let file_path = concat!("data/", stringify!($tname), ".nix");
                let contents = include_str!(concat!("data/", stringify!($tname), ".nix"));

                let vfs = ReadOnlyVfs::singleton(file_path, contents.as_bytes());

                let session = $sess;

                let mut buffer = Vec::new();
                vfs.iter().map(|entry| lint::lint(entry, &session)).for_each(|r| {
                    buffer.write(&r, &vfs, OutFormat::StdErr).unwrap();
                });

                let stripped = strip_ansi_escapes::strip(&buffer).unwrap();
                let out =  std::str::from_utf8(&stripped).unwrap();
                insta::assert_snapshot!(&out);
            }
        };
    }
}

test_lint! {
    bool_comparison,
    empty_let_in,
    manual_inherit,
    manual_inherit_from,
    legacy_let_syntax,
    collapsible_let_in,
    eta_reduction,
    useless_parens,
    empty_pattern,
    redundant_pattern_bind,
    unquoted_uri,
    empty_inherit,
    faster_groupby => session_info!("2.5"),
    faster_zipattrswith => session_info!("2.6"),
    deprecated_to_path => session_info!("2.4"),
    bool_simplification,
    useless_has_attr,
    repeated_keys,
    empty_list_concat
}
