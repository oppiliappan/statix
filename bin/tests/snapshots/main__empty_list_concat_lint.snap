---
source: bin/tests/main.rs
expression: "& out"
---
[W23] Warning: Unnecessary concatenation with empty list
   ╭─[data/empty_list_concat.nix:6:4]
   │
 6 │   ([] ++ [1 2 3])
   ·    ──────┬──────  
   ·          ╰──────── Concatenation with the empty list, [], is a no-op
───╯
[W23] Warning: Unnecessary concatenation with empty list
   ╭─[data/empty_list_concat.nix:9:4]
   │
 9 │   ([1 2 3] ++ [])
   ·    ──────┬──────  
   ·          ╰──────── Concatenation with the empty list, [], is a no-op
───╯
[W23] Warning: Unnecessary concatenation with empty list
    ╭─[data/empty_list_concat.nix:12:4]
    │
 12 │   ([] ++ [])
    ·    ────┬───  
    ·        ╰───── Concatenation with the empty list, [], is a no-op
────╯
[W23] Warning: Unnecessary concatenation with empty list
    ╭─[data/empty_list_concat.nix:15:4]
    │
 15 │   ([] ++ [] ++ [])
    ·    ───────┬──────  
    ·           ╰──────── Concatenation with the empty list, [], is a no-op
────╯
[W23] Warning: Unnecessary concatenation with empty list
    ╭─[data/empty_list_concat.nix:15:10]
    │
 15 │   ([] ++ [] ++ [])
    ·          ────┬───  
    ·              ╰───── Concatenation with the empty list, [], is a no-op
────╯
